api/src/main/java/org/apache/iceberg/io/SupportsRecoveryOperations.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+/**
+ * This interface is intended as an extension for FileIO implementations to provide additional
+ * best-effort recovery operations that can be useful for repairing corrupted tables where there are
+ * reachable files missing from disk. (e.g. a live manifest points to data file entry which no
+ * longer exists on disk)
+ */
+public interface SupportsRecoveryOperations {
+
+  /**
+   * Perform a best-effort recovery of a file at a given path
```

Comment:
How recovery is performed is left up to implementations but I think the one thing this interface is prescriptive on is that the expectation that it is *best effort*. It's not guaranteed that just because FileIO implementations implement the mixin that they will recover the file no matter what (e.g. there's a limit to versioning etc); as a result the interface doesn't surface an exception, rather just a boolean indicating success or not.

I guess we may want to qualify what it means for a file to be recovered (it may be obvious but since it's an interface, it's better to be clear):

It means that assuming no concurrent deletes of a file at that path, `fileIO.inputFile(path).exists()` must be true after this method call.

If that definition makes sense to folks: I can update @RussellSpitzer @danielcweeks

Reply:
Other thought I had:

We could offer another method which uplevels the manifestFile and attempts to recover the manifest and if that's successful, attempt recovery of the data files. I discarded this case since someone could just traverse the metadata tree on their own and just use `recoverFile(String path)`. Also in that API, probably would need a response that's higher fidelity (the simplest is Map<String, Boolean>) than just a boolean since maybe a caller wants to take action if the manifest was recovered but data files weren't etc.

Reply:
I would want to add to the notes that implementations should do their best not to throw any runtime exceptions and that this method is essentially assumed to be a no-op on false or something like that.

I think we also need to describe how this method will be used. For example, "When reading a file if any exception is thrown this method will be called in an attempt to remedy the exception."

Although if we use that description should we also passthrough a Throwable? 

A bigger question I have is whether or not this is needed as part of the general API or if it should be part of a Repair action that we talked about before.

Reply:
That's an interesting point @RussellSpitzer, I think initially we were planning to use this in something like the [proposed repair manifests](https://github.com/apache/iceberg/pull/10445) procedure.  I would be a little concerned about putting it in the normal read path because it could hide more serious issues if we're trying to repair table state on the fly.

Reply:
Agree on the comment that we should note that impls should do their best to avoid throwing runtime exceptions.

>I think we also need to describe how this method will be used. For example, "When reading a file if any exception is thrown this method will be called in an attempt to remedy the exception."

I actually don't think this should be recommended for the read path like @danielcweeks said performing recovery's on the fly probably will mask that the table was corrupted in the first place; I feel a user should always determine that their table is corrupted and then explicitly opt-in to recover the file.

>A bigger question I have is whether or not this is needed as part of the general API or if it should be part of a Repair action that we talked about before.

On the repair action, it's true that functions could technically reside specifically in the action. But the challenge is the actual implementations will look different for different FileIOs which is what is driving defining an interface.

To me even if the interface itself is used primarily in the context of repair actions, there's enough merit to generalizing the interface in the public API because different FileIO's will have different ways of recovering a given file so we probably want to standardize the method for that.